File: data_creation/simulation/src/Snake.cpp
#include "Snake.h"




Snake::Snake(int x, int y) {
    head = make_pair(x, y);
}

pair<int, int> Snake::get_move(char move) {
    switch (move) {
        case 'U':
            return make_pair(-1, 0);
        case 'D':
            return make_pair(1, 0);
        case 'L':
            return make_pair(0, -1);
        case 'R':
            return make_pair(0, 1);
        default:
            return make_pair(0, 0);
    }
}


void Snake::move_snake(char direction){
    pair<int, int> move_pair = get_move(direction);
    // cout << move_pair.first << endl;
    pair<int, int> head_cpy = head;

    head.first += move_pair.first;
    head.second += move_pair.second;

    if (tail.size() > 0){
        // get rid of the last segment
        tail.pop_back();
        // at the beginning of the tail insert the head
        tail.insert(tail.begin(), head_cpy);
    }

    // cout << head.first << endl;
}

pair<int, int> Snake::get_last_snake_segment(){
    if (tail.size() == 0){
        return make_pair(head.first, head.second);
    }
    return make_pair(tail.back().first, tail.back().second);
}

File: src/snake_game/snake.py
from typing import List, Tuple
import pygame

from .game_config import GameConfig

class Snake:
    def __init__(self, x: int, y: int):
       self.head: Tuple[int, int] = (x, y)
       self.tail: List[Tuple[int, int]] = []
       self.moves_history: List[Tuple[int, int]]  = []


    def get_move(self, direction: str):
       if direction == "W":
          return (-1, 0)
       elif direction == "S":
          return (1, 0)
       elif direction == "A":
          return (0, -1)
       elif direction == "D":
          return (0, 1)
       else:
          return (0, 0)
       
    def move(self, direction: str):
       move = self.get_move(direction)
      #  if move == (0,0):
      #     return
       head_cpy = self.head
       self.head = (self.head[0] + move[0], self.head[1] + move[1])

       # if snake collides with it
       
       if len(self.tail) > 0:
         self.tail.pop()
         self.tail.insert(0, head_cpy)

    def get_last_snake_segment(self):
         if len(self.tail) > 0:
          return self.tail[-1]
         else:
          return self.head


    